dcos/dcos-ui
# DCOS-57088 Force Update for Quota limits

This PR introduces the "Force Update" for updating quota on a group. Force Update is needed if you try to update a group's quota limits to lower than what is currently consumed by that group.

As a part of this work I also implemented the "live validation" for the group form after you've tried to save the first time. This is so that "force" mode would clear if you change anything in the form.

I also found and fixed a bug where "Edit Group" was not available in the header actions of the Quota View for top-level groups like it is for the Services tab.

## Testing
[DCOS-57088](https://jira.mesosphere.com/browse/DCOS-57088)

~** You need a cluster running the latest Mesos to test overcommitting see JIRA for links to DC/OS PR to deploy a cluster from~

1. Create a group with quota
(1.0 CPU, 510 Mem)
2. Create a service that will use most of the Quota
- `sleep infinity` with 5 instances @ 0.2 CPU & 100 Mem
3. Wait for service to deploy
4. Open edit group modal
5. Change Quota limits to lower than consumed values
(0.5 CPU, 250 Mem)
6. Verify:
- Top-level infobox error is displayed
- field level error is displayed
- Update button is now "Force Update" and `danger` style
7. Click "Force Update"
8. Verify Group quota was updated.

Notes changing any values when "Force Update" is present will intentionally clear the error.

## Trade-offs

I don't love how I add the `data?` field to the `GroupMutationResponse` but I also wasn't sure about a better option.

## Dependencies


This can be merged before Mesos lands in DC/OS you just won't see the "Force Update" workflow until Mesos quota validation errors are returned when overcommiting. 

## Screenshots


![image](https://user-images.githubusercontent.com/1972968/62979482-8a7d2e00-bde9-11e9-9a68-d68a94568bc1.png)
